internal/service/alarms/internal/db/repo/alarms_repository.go

```diff
@@ -51,3 +60,201 @@ func (ar *AlarmsRepository) GetAlarmEventRecordWithUuid(ctx context.Context, uui
 	slog.Info("AlarmEventRecord found", "uuid", uuid)
 	return []models.AlarmEventRecord{record}, nil
 }
+
+// UpsertAlarmEventRecord insert and updating an AlarmEventRecord.
+func (ar *AlarmsRepository) UpsertAlarmEventRecord(ctx context.Context, records []models.AlarmEventRecord) error {
+	// Build queries for each record
+	sql, params, err := buildAlarmEventRecordUpsertQuery(records, models.AlarmEventRecordView.Name(ctx).String())
+	if err != nil {
+		return fmt.Errorf("failed to build query for event upsert: %w", err)
+	}
+
+	_, err = ar.Db.Exec(ctx, sql, params...)
+	if err != nil {
+		return fmt.Errorf("failed to execute upsert query: %w", err)
+	}
+
+	slog.Info("Successfully inserted alerts from alertmanager", "table", models.AlarmEventRecordView.Name(ctx).String(), "count", len(records))
+	return nil
+}
+
+// buildAlarmEventRecordUpsertQuery builds the query for insert and updating an AlarmEventRecord
+func buildAlarmEventRecordUpsertQuery(records []models.AlarmEventRecord, tableName string) (string, []any, error) {
+	query := psql.Insert(im.Into(tableName))
+
+	// Dynamically construct the column list
+	columns := []string{
+		"alarm_raised_time", "alarm_changed_time", "alarm_cleared_time",
+		"alarm_acknowledged_time", "alarm_acknowledged", "perceived_severity",
+		"extensions", "resource_id", "resource_type_id",
+		"alarm_status", "fingerprint", "alarm_definition_id",
+		"probable_cause_id",
+	}
+
+	// Check if alarm_event_record_id is needed
+	hasAlarmEventRecordID := false
+	for _, record := range records {
+		if record.AlarmEventRecordID != nil {
+			hasAlarmEventRecordID = true
+			columns = append([]string{"alarm_event_record_id"}, columns...)
+			break
+		}
+	}
+
+	// Set the column list for the InsertQuery
+	query.Expression.Columns = columns
+
+	// Add the values for each record
+	values := make([]bob.Mod[*dialect.InsertQuery], 0, len(records))
+	for _, record := range records {
+		// Build the values for the row
+		args := []any{
+			record.AlarmRaisedTime, record.AlarmChangedTime, record.AlarmClearedTime,
+			record.AlarmAcknowledgedTime, record.AlarmAcknowledged, record.PerceivedSeverity,
+			record.Extensions, record.ResourceID, record.ResourceTypeID,
+			record.AlarmStatus, record.Fingerprint, record.AlarmDefinitionID,
+			record.ProbableCauseID,
+		}
+
+		if hasAlarmEventRecordID {
+			if record.AlarmEventRecordID != nil {
+				args = append([]any{record.AlarmEventRecordID}, args...)
+			} else {
+				args = append([]any{nil}, args...)
+			}
+		}
+
+		values = append(values, im.Values(psql.Arg(args...)))
+	}
+
+	// Apply the values to the query
+	query.Apply(values...)
+
+	// constraints only applicable for non-archive table
+	if tableName == models.AlarmEventRecordView.Name(context.Background()).String() {
+		query.Apply(im.OnConflictOnConstraint("unique_fingerprint_alarm_raised_time").DoUpdate(
+			im.SetExcluded("alarm_changed_time"),
```

Comment:
say we have an alert with a valid resourceID but we failed to collect the alarm dict, I suppose we will still report it to the user. Then, before the next notification comes, we manage to collect the alarm dict. We will update the alarm defintion ID which also means that we would have to update the alarm_changed_time since one of the values has changed, makes sense?

Reply:
alarm_changed_time is currently not being added since alertmanager is not sending it

Reply:
Isn't there an updatedAt field in the alert ?

Reply:
Hey Brent yeah it does but only if we call the API (pull). The notification doesnt have it https://prometheus.io/docs/alerting/latest/configuration/#webhook_config

Reply:
https://github.com/prometheus/alertmanager/ issues/2142#issuecomment-567087632 -- more info here